From b1638726aa726c23c775837f061454c60396133d Mon Sep 17 00:00:00 2001
From: Vinu Vaghasia <vinu.vaghasia@amd.com>
Date: Wed, 4 May 2022 16:48:39 -0500
Subject: [PATCH 1/1] systemd: Modify code to check for customized hostname
 setup

- Added code to set the hostname from bootargs.
- If the hostname is changed, then also update the /etc/hostname file
accordingly

Signed-off-by: Mohsen Dolaty <mohsen.dolaty@amd.com>
Signed-off-by: Vinu Vaghasia <vinu.vaghasia@amd.com>
---
 src/core/hostname-setup.c | 97 ++++++++++++++++++++++++---------------
 1 file changed, 59 insertions(+), 38 deletions(-)

diff --git a/src/core/hostname-setup.c b/src/core/hostname-setup.c
index 867ea19905..c39ebde7c7 100644
--- a/src/core/hostname-setup.c
+++ b/src/core/hostname-setup.c
@@ -14,50 +14,71 @@
 #include "string-util.h"
 #include "util.h"

+#define HOSTNAME_CMD_LEN    256
+
 int hostname_setup(void) {
-        _cleanup_free_ char *b = NULL;
-        const char *hn = NULL;
-        bool enoent = false;
-        int r;
-
-        r = proc_cmdline_get_key("systemd.hostname", 0, &b);
-        if (r < 0)
-                log_warning_errno(r, "Failed to retrieve system hostname from kernel command line, ignoring: %m");
-        else if (r > 0) {
-                if (hostname_is_valid(b, true))
-                        hn = b;
-                else  {
-                        log_warning("Hostname specified on kernel command line is invalid, ignoring: %s", b);
-                        b = mfree(b);
-                }
-        }
+    _cleanup_free_ char *b = NULL;
+    const char *hn = NULL;
+    bool enoent = false;
+    int r;
+    const char *etc_hn = NULL;
+    const char *cmdl_hn = NULL;
+    char hn_cmd_buff[HOSTNAME_CMD_LEN];

-        if (!hn) {
-                r = read_etc_hostname(NULL, &b);
-                if (r < 0) {
-                        if (r == -ENOENT)
-                                enoent = true;
-                        else
-                                log_warning_errno(r, "Failed to read configured hostname: %m");
-                } else
-                        hn = b;
+    /* Read hostname from bootargs */
+    r = proc_cmdline_get_key("systemd.hostname", 0, &b);
+    if (r < 0)
+            log_warning_errno(r, "Failed to retrieve system hostname from kernel command line, ignoring: %m");
+    else {
+        if (hostname_is_valid(b, true))
+            cmdl_hn = b;
+        else  {
+            log_warning("Hostname specified on kernel command line is invalid, ignoring: %s", b);
+            b = mfree(b);
         }
+    }

-        if (isempty(hn)) {
-                /* Don't override the hostname if it is already set and not explicitly configured */
-                if (hostname_is_set())
-                        return 0;
+    /* Read hostname from /etc/hostname file */
+    r = read_etc_hostname(NULL, &b);
+    if (r < 0) {
+        if (r == -ENOENT)
+            enoent = true;
+        else
+            log_warning_errno(r, "Failed to read configured hostname: %m");
+    }
+    else {
+        etc_hn = b;
+    }

-                if (enoent)
-                        log_info("No hostname configured.");
+    /* Set hostname from /etc/hostname file */
+    hn = etc_hn;

-                hn = FALLBACK_HOSTNAME;
-        }
+    /* check hostname from bootargs */
+    if ( (cmdl_hn != NULL) ) {
+        /* Set hostname from bootargs */
+        hn = cmdl_hn;
+        /* Update hostname to /etc/hostname file */
+        sprintf(hn_cmd_buff, "echo %s > /etc/hostname", hn);
+        r = system(hn_cmd_buff);
+        if ( r < 0)
+            return log_warning_errno(r, "Failed writing <%s> to /etc/hostname file: %m", hn);
+    }
+
+    if (isempty(hn)) {
+        /* Don't override the hostname if it is already set and not explicitly configured */
+        if (hostname_is_set())
+            return 0;
+
+        if (enoent)
+            log_info("No hostname configured.");
+
+        hn = FALLBACK_HOSTNAME;
+    }

-        r = sethostname_idempotent(hn);
-        if (r < 0)
-                return log_warning_errno(r, "Failed to set hostname to <%s>: %m", hn);
+    r = sethostname_idempotent(hn);
+    if (r < 0)
+        return log_warning_errno(r, "Failed to set hostname to <%s>: %m", hn);

-        log_info("Set hostname to <%s>.", hn);
-        return 0;
+    log_info("Set hostname to <%s>.", hn);
+    return 0;
 }
--
2.25.1

